[JSC] Add SensitiveToNaN flag to B3 Kind

Constellation opened the pull request:
#### 919ed00f75d8752cddfdacb59d286f1fc168a307
<pre>
[JSC] Add SensitiveToNaN flag to B3 Kind
<a href="https://bugs.webkit.org/show_bug.cgi?id=249991">https://bugs.webkit.org/show_bug.cgi?id=249991</a>
rdar://103807651

Reviewed by Saam Barati.

This patch adds SensitiveToNaN flag to B3 Kind. This means &quot;The B3 Value is sensitive to NaN bit pattern&quot;,
and it supresses some of strength reduction which can be observable as a NaN bit pattern difference.
From JS, this flag is always false since JS does not care about NaN bit pattern. On the other hand, Wasm
cares. In B3 Wasm IR generator, we put this flag to Wasm generated B3 Values.
This makes float spec-tests pass by 100%.

* JSTests/wasm/spec-tests/float_exprs.wast.js:
* Source/JavaScriptCore/b3/B3Kind.cpp:
(JSC::B3::Kind::dump const):
* Source/JavaScriptCore/b3/B3Kind.h:
(JSC::B3::Kind::Kind):
(JSC::B3::Kind::hasSensitiveToNaN):
(JSC::B3::Kind::hasSensitiveToNaN const):
(JSC::B3::Kind::sensitiveToNaN const):
(JSC::B3::Kind::setSensitiveToNaN):
(JSC::B3::Kind::operator== const):
(JSC::B3::Kind::hash const):
(JSC::B3::sensitiveToNaN):
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/wasm/generateWasmB3IRGeneratorInlinesHeader.py:
(CodeGenerator.generateB3OpCode):
(CodeGenerator):
(CodeGenerator.generateConstCode):
(CodeGenerator.generateOpcode):
(CodeGenerator.makeResult):
(CodeGenerator.generate):
(temp):
(generateB3OpCode): Deleted.
(generateConstCode): Deleted.

Canonical link: <a href="https://commits.webkit.org/258455@main">https://commits.webkit.org/258455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a01253701a83d2da566cbf5dd8b10d3c1c022a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34866 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111130 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171336 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11903 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1856 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94203 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108891 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107574 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9102 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92365 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/36829 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90992 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78667 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92206 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4538 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25281 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88358 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2137 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4618 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1726 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29403 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10703 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44768 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91266 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6369 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20394 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3055 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/JavaScriptCore/b3/B3Kind.h:
```diff
@@ -148,6 +146,35 @@ class Kind {
         m_traps = traps;
     }
+
+    static constexpr bool hasSensitiveToNaN(Opcode opcode)
```
"has" reads a bit poorly as a prefix here. Unless this is convention, I would go with "isSensitiveToNaN" or "supportsSensitiveToNaN" or the like. (might also be a good place for a comment explaining why these op codes are such).

Constellation replied:
Right now, this is a convention for these flags, like, we have hasTraps and hasIsChill.
For now, I'll just rename this flag to `isSensitiveToNaN`, and renaming this to `hasIsSensitiveToNaN`, aligned to the existing `hasIsChill`.

Constellation committed:
[JSC] Add SensitiveToNaN flag to B3 Kind
https://bugs.webkit.org/show_bug.cgi?id=249991
rdar://103807651

Reviewed by Saam Barati.

This patch adds SensitiveToNaN flag to B3 Kind. This means "The B3 Value is sensitive to NaN bit pattern",
and it supresses some of strength reduction which can be observable as a NaN bit pattern difference.
From JS, this flag is always false since JS does not care about NaN bit pattern. On the other hand, Wasm
cares. In B3 Wasm IR generator, we put this flag to Wasm generated B3 Values.
This makes float spec-tests pass by 100%.

* JSTests/wasm/spec-tests/float_exprs.wast.js:
* Source/JavaScriptCore/b3/B3Kind.cpp:
(JSC::B3::Kind::dump const):
* Source/JavaScriptCore/b3/B3Kind.h:
(JSC::B3::Kind::Kind):
(JSC::B3::Kind::hasSensitiveToNaN):
(JSC::B3::Kind::hasSensitiveToNaN const):
(JSC::B3::Kind::sensitiveToNaN const):
(JSC::B3::Kind::setSensitiveToNaN):
(JSC::B3::Kind::operator== const):
(JSC::B3::Kind::hash const):
(JSC::B3::sensitiveToNaN):
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/wasm/generateWasmB3IRGeneratorInlinesHeader.py:
(CodeGenerator.generateB3OpCode):
(CodeGenerator):
(CodeGenerator.generateConstCode):
(CodeGenerator.generateOpcode):
(CodeGenerator.makeResult):
(CodeGenerator.generate):
(temp):
(generateB3OpCode): Deleted.
(generateConstCode): Deleted.

Canonical link: https://commits.webkit.org/258455@main
